lib/argonaut/reservations.ex

```diff
@@ -0,0 +1,241 @@
+        inner_join: reservation in assoc(reminder, :reservation)
+      )
+    )
+  end
+
+  def everything do
+    from(
+      reservation in Reservation,
+      left_join: user in assoc(reservation, :user),
+      left_join: environment in assoc(reservation, :environment),
+      left_join: application in assoc(reservation, :application),
+      left_join: reminder in assoc(reservation, :reminder),
+      preload: [
+        user: user,
+        application: application,
+        environment: environment,
+        reminder: reminder
+      ]
+    )
+  end
+
+  def reservations_for_team(team_id) do
+    from(
+      reservation in everything(),
+      where: reservation.team_id == ^team_id
+    )
+  end
+
+  def reservations_by_user(user_id) do
+    from(
+      reservation in Reservation,
+      where: reservation.user_id == ^user_id,
+      join: environment in assoc(reservation, :environment),
+      join: application in assoc(reservation, :application),
+      preload: [application: application, environment: environment]
+    )
+  end
+
+  def reservation_with_associations(reservation_id) do
+    query =
+      from(
+        r in Reservation,
+        where: r.id == ^reservation_id,
+        join: a in assoc(r, :application),
+        join: e in assoc(r, :environment),
+        join: u in assoc(r, :user),
+        preload: [application: a, environment: e, user: u]
+      )
+
+    query |> Repo.one()
+  end
+
+  # TODO: These are copied from team_controller.ex. Keep them in one place
+  # Actions
+  # create_reservation and delete_reservation: These are different than the
+  # ones implemented in the channel file in that they take application name and
+  # environment name instead of ids now that an environment can be uniquely
+  # named and owned by one team, it will be easy to make reservations using
+  # just three pieces of info: user, application, environment
+  def create_reservation(%{"app" => app, "env" => env}, current_user) do
+    environment = Repo.one(from(env in Environment, where: env.name == ^env))
+
+    application =
+      Repo.one(
+        from(
+          app in Application,
+          where: app.name == ^app,
+          where: app.team_id == ^environment.team_id
+        )
+      )
+
+    {status, reason} =
+      satisfies?("", [
+        fn _ ->
+          if environment == nil, do: {:error, "No such environment #{env}"}, else: {:ok, ""}
+        end,
+        fn _ ->
+          if application == nil, do: {:error, "No such application #{app}"}, else: {:ok, ""}
+        end,
+        fn _ ->
+          if user_member_of_team?(current_user.id, environment.team_id),
+            do: {:ok, ""},
+            else: {:error, "None of the teams you are member of own the environment #{env}"}
+        end,
+        fn _ ->
+          if decide_and_reserve!(environment, application, current_user),
+            do: {:ok, ""},
+            else: {:error, "Someone else is using that environment currently"}
+        end
+      ])
```

Comment:
That's an interesting way to do that. Personally, I'd use a with chain here with explicitly defined functions for each of these cases. Should be a lot cleaner, more explicit, and concise.

See: https://www.openmymind.net/Elixirs-With-Statement/

Reply:
Yeah now I know how to use `with`, this is a copy pasta. I am not sure if I will get time to refactor this for this PR, but I will keep in mind for future changes.